doc/dev/code-generator.md

```diff
@@ -0,0 +1,211 @@
+- _parent key_: The parent key is given in the command line and defines the input keyset (the keys below the parent key).
+- _output name_: The base name for the output files as given in the command line. Suffixes may be appended, if there are multiple output
+  files.
+- _parts_: Each template may consist of multiple parts. Each part corresponds to exactly one mustache file and therefore exactly one output
+  file.
+- _part suffix_: Parts are identified by there part suffix. This suffix is appended to the template base name to obtain the name of the
+  mustache file corresponding to the part. Similarly it is appended to the output name to obtain the output file for this part. The empty
+  string `""` is a valid part suffix, although it is only recommend for templates with a single output file.
+
+## Creating a new template
+
+In this guide we will create a basic template, that generates a single file containing. The file will simple list all keys in our input
+keyset. An example output would be (the same as `kdb ls`):
+
+```
+user/sw/myapp/#0/current
+user/sw/myapp/#0/current/dir0
+user/sw/myapp/#0/current/dir0/subdir0/key0
+user/sw/myapp/#0/current/dir0/subdir0/key1
+user/sw/myapp/#0/current/dir1
+```
+
+### Creating the mustache template
+
+The first thing we need to create is a mustache template. You will need (at least) one template file for each output file. If you make use
+of partials (see [below](#using-partials)), you of course have to write multiple files.
+
+All template files must be located in the [`src/tools/kdb/gen/templates`](/src/tools/kdb/gen/templates) folder and must be named according
+to the scheme `<template base name>.<part suffix>.mustache`.
+
+For our simple example we create the file `src/tools/kdb/gen/templates/example.txt.mustache` with the following content:
+
+```
+{{# keys }}
+{{{ name }}}
+{{/ keys }}
+```
+
+Note: we will not go into detail on how mustache templates work, for more information see e.g.
+[here](https://mustache.github.io/mustache.5.html). All of features supported by the kainjow library should be supported by our framework
+as well.
+
+### Creating the supporting class
+
+Since we need to some way of supplying data to our template, we have to create a subclass of `GenTemplate`.
+
+First we create a new directory for our template class in `src/tools/kdb/gen`, for our template it should be `src/tools/kdb/gen/example`. In
+this directory we then create `example.cpp` and `example.hpp`. If your template becomes sufficiently complex, it may make sense to split it
+into multiple files, for this reason we recommend creating a new directory for each template.
+
+In `example.cpp` and `example.hpp` we create our subclass of `GenTemplate`. Therefore `example.hpp` should look like this:
+
+```cpp
+#ifndef ELEKTRA_EXAMPLE_HPP // choose a unique header guard
+#define ELEKTRA_EXAMPLE_HPP
+
+#include "../template.hpp"
```

Comment:
Relative includes should be avoided.

Reply:
Is there any particular reason for that?

Reply:
It is more clear if the different include files are in the same directory. Isn't this possible in this situation?

Reply:
It should be possible, I will update the guide and move the header of the existing template, when I rename it.

Reply:
I just saw that `src/tools/kdb` part of the include path, so I changed it to `#include <gen/template.hpp>`, which is a better solution anyway.